flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -61,6 +80,116 @@ static void transferAllStateDataToDirectory(
 		downloadDataForAllStateHandles(miscFiles, dest, restoringThreadNum, closeableRegistry);
 	}
 
+	/**
+	 * Upload all the files to checkpoint fileSystem using specified number of threads.
+	 *
+	 * @param files The files will be uploaded to checkpoint filesystem.
+	 * @param numberOfSnapshottingThreads The number of threads used to upload the files.
+	 * @param checkpointStreamFactory The checkpoint streamFactory used to create outputstream.
+	 * @param closeableRegistry
+	 *
+	 * @throws Exception Thrown if can not upload all the files.
+	 */
+	public static Map<StateHandleID, StreamStateHandle> uploadFilesToCheckpointFs(
+		@Nonnull Map<StateHandleID, Path> files,
+		int numberOfSnapshottingThreads,
+		CheckpointStreamFactory checkpointStreamFactory,
+		CloseableRegistry closeableRegistry) throws Exception {
+
+		Map<StateHandleID, StreamStateHandle> handles = new HashMap<>();
+
+		ExecutorService executorService = createExecutorService(numberOfSnapshottingThreads);
```

Comment:
This line makes me wonder if it is a good idea that this is just a static utils class instead of a proper class. One advantage of executor services is that they don't need to recreate threads all the time. I agree that this method is not called in any hot loops, so the overhead factor might be not dramatic, but still I wonder why not to make this a proper class that creates a threadpool once and only shuts it down at the end of the object lifecycle. Any reason why you chose this way?

Reply:
@StefanRRichter I chose this way to align the download function `downloadDataForAllStateHandles`. 
I thought `uploadFilesToCheckpointFs` function may receive both the `sstFiles` and `miscFiles` as parameter, so we can reuse the executorService, but in that way, makes the function `uploadFilesToCheckpointFs` a little harder to reuse.

Reply:
@klion26 
The upload and download part are quite independent. The base class could contain the executor, number of threads, closable registry and close method (to shutdown the executor instead of shutdowning it every time). The  upload and download classes can extend it and share the executor. The downloader could be then registered with the closable registry. For upload, we can of course shutdown it immediately after restore is done. I agree that it would make sense to keep threads up and for code reuse as well.